### PR TITLE
Feature: Native Options in Netty

### LINKED ISSUE #2066: Ability to use native transport options with Netty
### Steps to Reproduce

1. Try to use `TCP_CORK` or `TCP_QUICKACK`
2. Can't
3. See below in *Example Application* for code example

### Expected Behaviour

It would be really cool to be able to use higher-level options for `epoll` and `kqueue`

### Actual Behaviour

Instead, they are seen as "unknown options" on a given channel, even if `use-native-transport` is `true`. This is because Micronaut currently looks for socket options on the regular [`ChannelOption`](https://netty.io/4.0/api/io/netty/channel/ChannelOption.html), instead of knowing to look on the implementation-specific ones (like, say, [`EpollChannelOption`](https://netty.io/4.0/api/io/netty/channel/epoll/EpollChannelOption.html) and [`KQueueChannelOption`](https://netty.io/4.1/api/io/netty/channel/kqueue/KQueueChannelOption.html).

### Environment Information

- **Operating System**: `macOS 10.14.6` and `Alpine Linux`
- **Micronaut Version:** `master`
- **JDK Version:**
   ```
   java version "1.8.0_221"
   Java(TM) SE Runtime Environment (build 1.8.0_221-b11)
   Java HotSpot(TM) 64-Bit Server VM (build 25.221-b11, mixed mode)
   ```

### Example Application

**`application.yml`**
```yaml
micronaut:
  server:
    netty:
      use-native-transport: true
      childOptions:
        tcpNodelay: true
        tcpQuickack: true
````

*Note: I'm told one should never enable both `TCP_NODELAY` and `TCP_QUICKACK` at the same time, but I'm doing it here to illustrate that one is supported, and the other is not.*

Here is the output when trying to run an app with the above config:
```
17:25:25 [INFO]: Startup completed in 2152ms. Server Running: http://0.0.0.0:8080
17:25:40 [WARN]: Unknown channel option 'TCP_QUICKACK' for channel '[id: 0x23d15fe9, L:/0:0:0:0:0:0:0:1%0:8080 - R:/0:0:0:0:0:0:0:1%0:63150]'
[... app logs begin...]
```

### PR DESCRIPTION
**Pull request by sgammon**
This changeset adds the ability to set options supported by native Netty-supported transports, like `epoll` and `kqueue`.

Micronaut already makes it easy to specify `options` and `childOptions` under Netty, but those options don't offer support for higher-level stuff exclusively supported by transports like `epoll` and `kqueue`.

- Users in certain circumstances may want to turn on `TCP_NODELAY`.   This is already easy via Micronaut's support for Netty options,   because `TCP_NODELAY` is part of `io.netty.channel.ChannelOption`,   which is where Micronaut looks for socket options.

- Users on Linux might want to take John Nagle's advice and, instead of enabling `TCP_NODELAY`, enabling the more modern `TCP_QUICKACK` option (supported since Linux 2.4.4). This can't be done out of the box in Micronaut, however, because that option is defined in Netty at `io.netty.channel.epoll.EpollChannelOption`.

This fixes and closes #2066, which I filed to track the issue. Basically, `ChildOption` instances in the `NettyHttpServerConfiguration` are now wrapped in an object that describes whether they are base options, or options for specific transports.

This enables a few new config options that look like this:
```yaml
micronaut:
  server:
    netty:
      childOptions:
        soKeepalive: true
      epoll:
        childOptions:
          tcpQuickack: true
      kqueue:
        childOptions:
          tcpNopush: true
```

At runtime, the transport is detected if `use-native-transport` is turned on. If it's `epoll` or `kqueue`, their individual options are merged into a combined map, with the newly-wrapped `ChildOption` instances. This enables `processOptions` to resolve the option to set on the appropriate class.

- [x] Add `NettyChannelOption` private wrapper class in `NettyHttpServerConfiguration`
- [x] When gathering options, wrap them in `NettyChannelOption`
- [x] When applying options, resolve them via `NettyChannelOption`
- [x] Declare new configuration options
- [ ] Add unit testing
- [ ] Testing downstream in production

### CODE REVIEW

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java [hunk 1]
```diff
@@ -542,23 +553,133 @@ private void registerMicronautChannelHandlers(ChannelPipeline pipeline) {
...
+                            break;
+                        case KQUEUE:
+                            optionTarget = KQueueChannelOption.class;
+                            break;
+                        default:
+                            throw new IllegalArgumentException(
+                              "Unrecognized Netty channel option type: " + channelOption.type.name());
+                    }
+
+                    Optional<Field> declaredField = ReflectionUtils.findDeclaredField(optionTarget, name);
```
[the remark this answers is not in the document]
Comment (sgammon) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java]: happy to figure that one out but i'll have to google it, if you have a link or code sample describing how it should be done, happy to update it as i go

[the remark this answers is not in the document]
Comment (jameskleeh) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java]: See https://docs.micronaut.io/latest/guide/index.html#introspection

**Comment by jameskleeh** [comment]
@sgammon Are you interested in moving this PR forward or should I close?

**Comment by sgammon** [comment]
@jameskleeh i haven't had a lot of time to pursue this lately, but i'll be picking it back up soon. if you'd rather close to keep things clean, that's fine, i can file on my branch and work there until it's ready.

**Comment by sgammon** [comment]
@jameskleeh i have a version working that applies the native options in question, but it doesn't appear to actually apply them on the underlying socket for some reason. i'm still working out if it's me, micronaut, or netty.

**Comment by jameskleeh** [comment]
@sgammon No worries, we'll keep it open here in case you want to bring items up for discussion

**Comment by graemerocher** [comment]
Replaced by #3056

**Pull request closed without merging**